feat(client): coming soon chapters, modules and blocks

ojeytonwilliams opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

If a block doesn't exist yet, we can now create that block in the superblock [structure](https://github.com/freeCodeCamp/freeCodeCamp/blob/main/curriculum/superblock-structure/full-stack.json) and it will render as a "coming soon" block on the superblock page.

For example, https://github.com/freeCodeCamp/freeCodeCamp/blob/ea15720fe34f134a54ee01c203ebd12426d4c41c/curriculum/superblock-structure/full-stack.json#L140-L146

would become

```json
          "dashedName": "css-colors",
          "blocks": [
            { "dashedName": "lecture-working-with-colors-in-css" },
            { "dashedName": "workshop-colored-markers" },
            { "dashedName": "review-css-colors" },
            { "dashedName": "this-block-is-not-ready-yet" },
            { "dashedName": "quiz-css-colors" }
          ]
```

and that module would look like this:

![new module ui](https://github.com/user-attachments/assets/1e236264-8074-4b01-86d8-3a5d2caaf995)

It's also possible to create entire empty chapters or modules in the same way. E.g.

```json
    {
      "dashedName": "coming-soon-chapter-1",
      "modules": [
        {
          "dashedName": "coming-soon-module-1",
          "blocks": [{ "dashedName": "coming-soon-block" }]
        }
      ]
    }
```

and that will look like this

![image](https://github.com/user-attachments/assets/6ecff7cd-fe8b-44b0-bf08-66afadd581cf)

Ref: https://github.com/freeCodeCamp/fcc10/issues/41 (arguably, it closes it, but the UI is horrid)

<!-- Feel free to add any additional description of changes below this line -->

ojeytonwilliams committed:
refactor: rename variables

They're not all the chapters/modules/blocks, just the current ones.

ojeytonwilliams committed:
refactor: derive expanded state from config

Uses the superblock structure to define the relationship between blocks
and chapters and modules. This is a little more direct and should make
it easier to expand currentChapters to include all chapters.

ojeytonwilliams committed:
feat(client): show coming soon blocks

moT01 commented:
Would we be able to work on that module while still showing it as "coming soon?" We will likely be in the middle of creating these, or have them partially finished at certain points, and still want to show them as coming soon until they are completed.

ojeytonwilliams commented:
I'll have to check, but it should work already. 

If a block is upcoming it won't have challenges and so will be rendered as "coming soon". 

If it is not upcoming, it'll get rendered as normal.

ojeytonwilliams committed:
fix: handle all missing blocks

ojeytonwilliams committed:
feat: show coming soon chapters and modules

ojeytonwilliams committed:
feat: use coming soon badge

ojeytonwilliams committed:
feat: use Coming Soon badge for chapters/modules

ojeytonwilliams committed:
fix: use right key

ojeytonwilliams committed:
feat: use Coming Soon badge for blocks

ojeytonwilliams committed:
fix: add coming soon translation

ojeytonwilliams committed:
refactor: remove ComingSoon blocklayout

It's not necessary now that the superblock accordion renders the badge
instead.

ojeytonwilliams commented:
![image](https://github.com/user-attachments/assets/a0d8c0da-05e2-44f7-88a2-07acece186cb)

It could use some work, but now all the chapters, modules and blocks that appear in the superblock structure, but are empty will be rendered with the coming soon badge.

That means if you want a coming soon chapter, do this:

```json
    {
      "dashedName": "name-of-new-chapter",
      "modules": []
    }
 ```

If you want a coming soon module:

```
    {
      "dashedName": "name-of-existing-chapter",
      "modules": [
        {
          "dashedName": "name-of-new-module",
          "blocks": []
        }
      ]
    },
```

and finally, if you want a coming soon block, just leave `isUpcomingChange: true` in that block's `meta.json`.

ojeytonwilliams committed:
fix(style): update coming soon styles

ojeytonwilliams committed:
refactor: reorganise components

ojeytonwilliams committed:
fix: overlapping border declarations

ojeytonwilliams committed:
fix: remove unused fallback

moT01 committed:
fix: move coming soon to full stack.json

moT01 commented:
Well, I don't know how much I love it, but it seems to be working pretty decent @ojeytonwilliams.

With showUpcomingChanges = true, you see all the challenges as usual.
With showUpcomingChanges = false, the fsd superblock page doesn't build because there's no challenges - which I think is the current behavior. So you need to set `isUpcomingChange: false` in one of the `meta.json` files. Set it in the `lecture-introduction-to-javascript-libraries-and-framework` block to see the front end libraries chapter - which is the one with react and the upcoming modules.

ojeytonwilliams committed:
refactor: give TS hints about the types

We don't need additional logic, just to tell TS what the type really is.

ojeytonwilliams committed:
refactor: handle empty blocks

ojeytonwilliams committed:
feat: render chapter/module unless coming soon

Even if a chapter/module is empty, it should be shown. The comingSoon
flag allows us to hide content, so we should use it.

ojeytonwilliams committed:
chore: mark all chapters as comingSoon

When we're happy with them, we can delete the flag, but we should
make that decision, not default to them being shown.

moT01 commented on curriculum/superblock-structure/full-stack.json:
```diff
@@ -664,6 +675,16 @@
+      "modules": []
+    },
     {
       "chapterType": "exam",
```
It really doesn't matter that much. It will stay hidden until we change `isUpcomingChange: false` on that block. I think it's fine as-is. The more important thing here is that it easily allows it to add any items as coming soon when we need to.

ojeytonwilliams replied:
That's a good point. Also, if you want to show it as coming soon, it's a trivial change. Add the flag and update intro.json and it'll pop into being.
